### ISSUE TITLE
TSSr-submission

### ISSUE THREAD

**Issue by Linlab-slu** [Bioconductor/Contributions#2222]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/Linlab-slu/TSSr

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [https://bioconductor.org/developers/package-submission/][2]
  instructions. My package is consistent with the _Bioconductor_
  [https://contributions.bioconductor.org/][1].

- [x] I understand Bioconductor [https://bioconductor.org/developers/package-submission/#naming][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [https://support.bioconductor.org][3] for issues that users may
  have, subscribing to the [https://stat.ethz.ch/mailman/listinfo/bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [https://bioconductor.org/about/code-of-conduct/][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [http://bioconductor.org/developers/how-to/git/][5]
  (optionally [http://bioconductor.org/developers/how-to/git/sync-existing-repositories/][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [https://bioconductor.org/][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by hpages** [comment]
Hi @Linlab-slu ,

Please address the errors and warnings reported by the SPB (Single Package Builder). Thanks!

Best,
H.

**Comment by hpages** [comment]
Hi @Linlab-slu ,

Thanks for fixing the errors. `BiocCheck("TSSr_0.99.3.tar.gz")` still reports many NOTEs: 14 in total, which is a lot! Many of them deserve some attention and should be easy to address. For example: 
```
    * NOTE: 'sessionInfo' not found in vignette(s)
    ...
    * NOTE: Avoid sapply(); use vapply()
    ...
    * NOTE: Avoid 1:...; use seq_len() or seq_along()
    ...
    * NOTE: Avoid redundant 'stop' and 'warn*' in signal conditions
    ...
    * NOTE: Consider adding runnable examples to the following man
      pages which document exported objects:
      callEnhancer.Rd, clusterTSS.Rd, getTSS.Rd, shapeCluster.Rd,
      shiftPromoter.Rd
```
Please reduce the number of NOTEs to <= 8.

Thanks,
H.

**Comment by hpages** [comment]
Hi @Linlab-slu ,

Were you able to take a look at this?

Thanks,
H.

**Comment by Linlab-slu** [comment]
Sorry for the late updates. Yes, I would definitely reduce some notes and repush as soon as possible. 
Thanks a lot!

**Comment by Linlab-slu** [comment]
@hpages 

Sorry for the late updates. Yes, I would definitely reduce some notes and repush as soon as possible.

Thanks a lot!

**Comment by hpages** [comment]
You've only addressed 1 NOTE out of the 14 NOTEs. Is this the best you can do? I've pointed you to other NOTEs that deserve your attention and should be easy to address. Thanks!

H.

**Comment by hpages** [comment]
Thanks for the improvements @Linlab-slu.

A few more things:

1. Why are the examples in the man pages for `callEnhancer()`,  `getTSS()`, and `shapeCluster()` commented out? Per our guidelines, all user facing functions need to have at least one working example.

2. Why does the `\value` section in the man page for `callEnhancer()` say that the function returns a "Large List of elements - one element for each sample" when this function seems to return a TSSr object? (Note that TSSr is an S4 class and TSSr object have _slots_, while an ordinary _list_ has _list elements_).

3. The TSSr object returned by `callEnhancer()` contains data.table objects (e.g. in the `PromoterShift` slot). Unfortunately, while it's fine to use data.table objects internally in your package, we discourage exposing these objects to the end user. This is because they don't obey the "standard data.frame API" e.g. `DT[2]` extracts the 2nd row when it should extract the 2nd column. It wouldn't be a big problem if data.table was not extending data.frame, but it does. Concretely this means that the huge amount of code around (in Bioconductor and CRAN packages) that was written in a generic fashion to work on objects for which `is.data.frame()` is TRUE is most likely going to choke on a data.table object. This is an unfortunate situation and I wish the data.table developers had stuck to the standard data.frame contract here. So please make sure to always present data.frame or DataFrame objects instead of data.table objects to the end user.

Thanks,
H.

**Comment by zhaolianlu** [comment]
Hi, @hpages 

Thanks for your comments! Here are my explanations regarding your concerns: 

1. These functions are the speed limit steps in CMD check, so I excluded them out for checking.
2. No. callEnhancer() function returns list of data tables back into the object in the . There is nothing to output or print externally when running callEnhancer() functions. 
3. That is a good point. We have many slots containing data.table objects internally. I checked the whole pipeline, and believed we don't expose data.table objects to the end user, except when the users type object name to see what's in it. That's should be fine from my sense. If the user want to access the results from some functions like callEnhancer(), we provided a list of functions, such as exportEnhancerTable(), to export the whole table into a file.

Thanks a lot! Please let me know if you still have any concerns.
ZL

**Comment by hpages** [comment]
Hi @zhaolianlu ,

Sorry for the delay.

**Item 1:**

> These functions are the speed limit steps in CMD check, so I excluded them out for checking.

But it takes only 0.007 s to run the `exportEnhancerTable` example on my laptop:
```
system.time(res2 <- exportEnhancerTable(exampleTSSr))
# Exporting enhancer table...
#    user  system elapsed 
#   0.004   0.000   0.007 
```
and only 2.18 s to run the `callEnhancer` example:
```
system.time(res <- callEnhancer(exampleTSSr,flanking = 400,dis2gene=2000))
# Calculating enhancers...
#    user  system elapsed 
#   2.171   0.048   2.180 
# Warning messages:
# 1: In `[.data.table`(cs, , `:=`(gene, NULL)) :
#   Column 'gene' does not exist to remove
# 2: In `[.data.table`(cs, , `:=`(inCoding, NULL)) :
#   Column 'inCoding' does not exist to remove
# 3: In `[.data.table`(cs, , `:=`(gene, NULL)) :
#   Column 'gene' does not exist to remove
# 4: In `[.data.table`(cs, , `:=`(inCoding, NULL)) :
#   Column 'inCoding' does not exist to remove
```
  Anyways, disabling examples is not the way to go as it almost guarantees that the code will rot at some point. We can already see some warnings when running the above code which suggests that the code rot has already started, maybe?
  Per our guidelines, all user-facing functions must be documented and have running examples. Please make sure to comply and also make sure to get rid of these warnings as they can only confuse the end user.

**Item 2:**

Here is how the value returned by `callEnhancer()` is described in the man page:

> Large List of elements - one element for each sample

However, what I get is a _TSSr object_, not a _list_:
```
> class(res)
[1] "TSSr"
attr(,"package")
[1] "TSSr"
```
So your description of the returned value is inaccurate and way too vague. What's exactly in this object, how to access its content, and, more generally, what to do with it, should be precisely documented.

**Item 3:**

As long as the end user never has to manipulate directly those data.table objects we should be fine.

One more thing:

**Item 4:**

It looks like you've taken the approach of making all your functions S4 generics, in a systematic way. There's no need for this and it's not considered good practice (it adds an unnecessary layer of complexity and makes it harder to debug the functions). Please use regular functions instead.

Thanks,
H.

**Comment by hpages** [comment]
Dear @zhaolianlu, do you intend to follow up with this submission? Thanks. H.

**Comment by lshep** [comment]
@zhaolianlu  When you are ready with changes and to continue with the review process please request that this issue be reopened.

**Comment by JohnnyChen1113** [comment]
Hello,

  I am a member of the Linlab team and would like to reactivate this submission.

  Since the original submission in 2021, we have made significant improvements to the package:

  **Updates in v0.99.6:**
  - Added unit tests (testthat framework, 6 test files)
  - Fixed NAMESPACE issues (removed hardcoded BSgenome import)
  - Fixed temp file handling (now uses `tempdir()`)
  - Fixed documentation typos
  - Updated CITATION format to `bibentry()`
  - Package now passes `R CMD check` with 0 ERROR, 0 WARNING, 0 NOTE

  **Current status:**
  - Repository: https://github.com/Linlab-slu/TSSr
  - Version: 0.99.6
  - All checks passing

  Could this submission be reactivated for review? We are committed to addressing any feedback promptly.

  Thank you!

**Comment by lshep** [comment]
@JohnnyChen1113 Can you clarify who is the maintainer and who needs push access to make changes to the repository at git.bioconductor.org?   As of right now only @Linlab-slu associated with maintainer email zhenguo.lin@slu.edu has push access to trigger a new build to continue the process.

**Comment by Linlab-slu** [comment]
> [@JohnnyChen1113](https://github.com/JohnnyChen1113) Can you clarify who is the maintainer and who needs push access to make changes to the repository at git.bioconductor.org? As of right now only [@Linlab-slu](https://github.com/Linlab-slu) associated with maintainer email [zhenguo.lin@slu.edu](mailto:zhenguo.lin@slu.edu) has push access to trigger a new build to continue the process.

I am the maintainer of this package. I will push the updates to git.bioconductor.org to trigger a new build. Please grant fetch/push permission to allow us to push new version to Bioconductor.

**Comment by lshep** [comment]
@Linlab-slu  you should activate your BiocCredntials account https://git.bioconductor.org/BiocCredentials/  associated with zhenguo.lin@slu.edu.  Once activated you can update ssh keys if necessary for push access.

**Comment by Linlab-slu** [comment]
> [@Linlab-slu](https://github.com/Linlab-slu) you should activate your BiocCredntials account https://git.bioconductor.org/BiocCredentials/ associated with [zhenguo.lin@slu.edu](mailto:zhenguo.lin@slu.edu). Once activated you can update ssh keys if necessary for push access.

Thank you! We have resolved the issue and succefully complete the push.

**Comment by lshep** [comment]
It did not appear to trigger a version bump.  Can you please bump the version in the package?

**Comment by JohnnyChen1113** [comment]
> Dear Package contributor,
> 
> This is the automated single package builder at bioconductor.org.
> 
> Your package has been built on the Bioconductor Single Package Builder.
> 
> On one or more platforms, the build results were: "WARNINGS". This may mean there is a problem with the package that you need to fix. Or it may mean that there is a problem with the build system itself.
> 
> Please see the [build report](http://bioconductor.org/spb_reports/TSSr_buildreport_20251213120357.html) for more details.
> 
> The following are build products from R CMD build on the Single Package Builder: Linux (Ubuntu 24.04.3 LTS): [TSSr_0.99.8.tar.gz](https://bioconductor.org/spb_reports/TSSr_0.99.8.tar.gz)
> 
> Links above active for 21 days.
> 
> ** Remember: **if you submitted your package after July 7th, 2020, when making changes to your repository push to `git@git.bioconductor.org:packages/TSSr` to trigger a new build. A quick tutorial for setting up remotes and pushing to upstream can be found [here](https://contributions.bioconductor.org/git-version-control.html#new-package-workflow).

Regarding the WARNING 'No Bioconductor dependencies detected': This appears to be a false positive. TSSr imports several Bioconductor packages including **GenomicRanges, rtracklayer, GenomicFeatures, DESeq2, Gviz, IRanges, Rsamtools, GenomicAlignments, GenomeInfoDb, and BiocGenerics** as listed in the DESCRIPTION file.
The NOTEs are coding style suggestions that we will address in future updates. @lshep

**Comment by JohnnyChen1113** [comment]
Hi @lshep ,
I wanted to follow up on the TSSr package submission. The latest build completed with WARNINGS status.

A few quick notes:

1. "No Bioconductor dependencies detected" - This appears to be a false positive. The DESCRIPTION file imports several Bioconductor packages including GenomicRanges, rtracklayer, GenomicFeatures, DESeq2, Gviz, IRanges, Rsamtools, GenomicAlignments, GenomeInfoDb, and BiocGenerics.
2. NOTEs regarding coding style - I'm happy to address these in a future update if needed.

Could you let me know if there are any blocking issues I should fix before the package can proceed to review? I'm aiming for the Spring 2026 release if possible.
Thank you for your time!

Best,
Junhao

**Comment by hpages** [comment]
Hi @JohnnyChen1113,

Thanks for the update and sorry for the slow response. No blocking issues. I'm going to take a look at the package today and will provide some feedback. Thanks for your patience.

H.

**Comment by hpages** [comment]
**The following `R CMD check` NOTE needs to be addressed:**

1. Undefined global functions or variables: `makeTxDbFromGFF`
   This means concretely that you need to import the `makeTxDbFromGFF` function. Note that the function is defined in the **txdbmaker** package (it was moved from **GenomicFeatures** to **txdbmaker** a couple of years ago).

**The following `BiocCheck()` NOTEs need to be addressed:**

2. Avoid `sapply()`; use `vapply()`.
   I will add here that sometimes it seems that what you really want to use is `lapply()`, not `sapply()` or `vapply()`, because further down you call `unlist()` on the result returned by `sapply()`. This means that you don't expect `sapply()` to be able to simplifiy anything, ever, which means that you shouldn't be using it in the first place. It's important that you understand the difference between `lapply()`, `sapply()` and `vapply()`. Please consult their man page for more information.

3. Avoid `cat` and `print` outside of `show` methods.
   For example, if the intention is to make a function a little bit verbose by displaying what it's currently doing, you should use:
    ```r
    message("process is running on", numCores, "cores...")
    ```
   instead of:
    ```r
    print(paste("process is running on", numCores, "cores..."))
    ```
   The benefits of `message()` over `print()` are: (a) the former produces cleaner output, and (2) the user can mute it with `suppressMessages()` if they wish.

4. Avoid using `=` for assignment and use `<-` instead.
   Clean R code should always use `<-` instead of `=` for assignment. This is considered good practise as it improves code readibility.

5. Consider adding runnable examples to man pages that document exported objects.
   Man pages that need examples: `callEnhancer.Rd`, `clusterTSS.Rd`, `exportClustersTable.Rd`, `exportClustersToBed.Rd`, `exportDETable.Rd`, `exportEnhancerTable.Rd`, `exportShapeTable.Rd`, `exportShiftTable.Rd`, `exportTSStable.Rd`, `exportTSStoBedgraph.Rd`, `getTSS.Rd`, `plotCorrelation.Rd`, `plotDE.Rd`, `plotInterQuantile.Rd`, `plotShape.Rd`, `plotTssPCA.Rd`, `plotTSS.Rd`, `shapeCluster.Rd`, `shiftPromoter.Rd`.
   Note that these man pages need _at least one example_ that illustrates the use of the function documented in the man page, and that example cannot be in a `\donttest{}` or `\dontrun{}` statement because it needs to be run by `R CMD check`. 

**Other things:**

6. As per Bioconductor package guidelines, direct use of `new()` by the user to create an object is not recommended. `new()` is too low-level and not user-friendly as it makes it too easy to create invalid objects. This is exacerbated by the fact that you didn't implement a validity method for the TSSr class. Bioconductor strongly suggests to provide a constructor function, with the convention to name the function like the class i.e. `TSSr()`. The function should perform sanity checks on the user input and make sure that the supplied values make sense, and fail graciously (i.e. with an informative error message) if they don't. Finally the constructor function needs to be documented (typically in the same man page where the class is documented) with examples provided.

7. Your TSSr objects have a design problem which is illustrated by the following code:
   ```r
   data(exampleTSSr)
   getTSS(exampleTSSr)
   Reading in file: S01.sorted.bam...
   # Error in value[[3L]](cond) : 
   #   failed to open BamFile: file(s) do not exist:
   #   'S01.sorted.bam'
   ```
   Disappointing! Would you agree with me that if a basic functionality like `getTSS()` doesn't work on this object then it's fair to say that the object is broken? Please include an object in your package that is fully functional. Maybe that means making the original BAM files available on ExperimentHub (via a separate data experiment package). If access to the BAM files is not a strong requirement for doing meaningful work with the object (that's for you to judge), then at the very least `getTSS()` should fail more graciously. Right now it really feels that something is broken.

8. The vignette has too many unevaluated code chunks (i.e. chunks where `eval=FALSE`).

9. The Format section in the man page for `exampleTSSr` says:
    ```
     An object of class TSSr
    
     exampleTSSr
          RangedSummarizedExperiment:TSSr
    ```
   What does `RangedSummarizedExperiment:TSSr` mean? What have TSSr objects to do with RangedSummarizedExperiment objects?

10. All the stuff in the `docs/` folder is static and was generated in 2021! How do we know it's still in sync with the content of the package? My bet is that it's probably not, and there's no easy way to know, which is why this kind of static content is _strongly_ discouraged. Please consider removing this folder.

11. The `Meta/` folder does not belong and should not be in `git`. This folder is generated and added to the package source tarball by `R CMD build`.

12. The package has 125 dependencies (direct and indirect). That's a lot! Are you sure that you need all the packages that are currently listed in `Imports`? For example, why is **rmarkdown** listed there? Where is it used and for what? Is **rmarkdown** required for normal operations of the package? It doesn't seem so. If it's only needed to build the vignette, then it should be listed in `Suggests` not in `Imports`.
   Another example is the **GenomicAlignments** package from which you import `qwidth`, `readGAlignments`, and `seqnames`. However the code in the package never calls `qwidth()` or `readGAlignments()`. It only calls `seqnames()`, which is a generic function defined in the **Seqinfo** package. So you should import `seqnames` from **Seqinfo** and get rid of the dependency on **GenomicAlignments**.
   And so on...
   Please take the time to revisit the package dependencies and make sure to import only what is actually needed by the internal code in the package (i.e. by the code in the `R/` folder).

**Finally:**
 
13. Please check the last feedback I sent before that issue was closed in December 2021 [here](https://github.com/Bioconductor/Contributions/issues/2222#issuecomment-983873686). Most of the items listed there have not been addressed.

Thanks,
H.

**Comment by lshep** [comment]
@JohnnyChen1113 do you plan to update the package based on the review or should we close this issue ?

**Comment by JohnnyChen1113** [comment]
> [@JohnnyChen1113](https://github.com/JohnnyChen1113) do you plan to update the package based on the review or should we close this issue ?

Hi @lshep, yes, we are still actively working on addressing the review feedback. We plan to submit an updated version soon. Thanks for following up!

**Comment by lshep** [comment]
@JohnnyChen1113  if there is not going to be an update soon we can close this for inactivity and reopen when you are ready with updates.  Is there an estimated time frame for updates?

**Comment by JohnnyChen1113** [comment]
> [@JohnnyChen1113](https://github.com/JohnnyChen1113) if there is not going to be an update soon we can close this for inactivity and reopen when you are ready with updates. Is there an estimated time frame for updates?

Hi @hpages,

Thank you for the thorough review on Feb 3. We've now addressed all the
items in TSSr 0.99.9. Summary of changes:

**R CMD check NOTE**
- Imported `makeTxDbFromGFF` from **txdbmaker** (NAMESPACE).

**BiocCheck NOTEs**
- Replaced all `sapply()` with `vapply()` or `lapply()` as appropriate.
- Replaced `cat()`/`print()` with `message()` outside of show methods.
- Switched all assignments from `=` to `<-` in R code.
- Added runnable examples (no `\donttest{}`/`\dontrun{}`) to all 18 man
  pages you listed: callEnhancer, clusterTSS, exportClustersTable,
  exportClustersToBed, exportDETable, exportEnhancerTable,
  exportShapeTable, exportShiftTable, exportTSStable, exportTSStoBedgraph,
  getTSS, plotCorrelation, plotDE, plotInterQuantile, plotShape,
  plotTssPCA, plotTSS, shapeCluster, shiftPromoter.

**Other items**
1. Added a `TSSr()` constructor with sanity checks on user input plus a
   validity method on the class. Both are documented with runnable examples.
2. `getTSS()` and related methods now check for the presence of input
   files and fail gracefully with an informative error when they are
   missing. We chose this path rather than shipping the BAM files in the
   package — a companion data experiment package (`TSSrData`) is in
   preparation and will be submitted separately to ExperimentHub once
   TSSr is accepted.
3. Reduced `eval=FALSE` chunks in the vignette to the minimum (only the
   chunks that require BAM-file inputs which are not bundled).
4. Rewrote the `Format` section of `exampleTSSr.Rd` to correctly describe
   the S4 TSSr class.
5. Removed the static `docs/` folder.
6. Removed the tracked `Meta/` folder; it is generated by `R CMD build`.
7. Trimmed `Imports`: moved `rmarkdown` to `Suggests`, dropped
   `GenomicAlignments`, and reviewed every remaining dependency to
   ensure each one is actually used by code under `R/`.

**Architectural item — request for guidance**

The 2021-12 feedback also asked us to remove the `setGeneric()` +
`assign(..., envir = parent.frame())` pattern. Refactoring this is a
substantial breaking change for downstream users of the API. We would
like to address it in a follow-up release rather than block this
submission. Could you let us know if that is acceptable?

**Local checks (R 4.6 / Bioc devel 3.23)**
- `R CMD check`: 0 ERROR / 0 WARNING.
- `BiocCheck`: 0 ERROR / 0 WARNING / 6 NOTES (all style or
  recommendation-level: function length, line width, biocViews
  suggestion, R version recommendation, function indent, mailing-list
  subscription detection).

The maintainer (@Linlab-slu) is subscribed to bioc-devel.

Thanks for your patience and for the very helpful review.

Best,
Junhao

**Comment by JohnnyChen1113** [comment]
> [@JohnnyChen1113](https://github.com/JohnnyChen1113) if there is not going to be an update soon we can close this for inactivity and reopen when you are ready with updates. Is there an estimated time frame for updates?

Hi @lshep,

We've prepared TSSr 0.99.9 with all of @hpages's Feb 3 review items
addressed (latest on GitHub: https://github.com/Linlab-slu/TSSr,
commit e469b74, tagged v0.99.9), but we're stuck pushing to
git.bioconductor.org.

The maintainer (Zhenguo Lin, BiocCredentials user `Linlab-slu`,
email `zhenguo.lin@slu.edu`) tried `git push upstream master` and got:

remote: FATAL: W refs/heads/master packages/TSSr Linlab-slu DENIED by fallthru
remote: error: hook declined to update refs/heads/master
! [remote rejected] master -> master (hook declined)

We've double-checked everything we can on our side:

- `ssh -T git@git.bioconductor.org` identifies him correctly as
  `Linlab-slu`.
- `ssh git@git.bioconductor.org info packages/TSSr` reports `W`
  permission for this user.
- BiocCredentials shows the maintainer email `zhenguo.lin@slu.edu`
  with the SSH key registered.
- The last successful push was Dec 13, 2025 (commit `49e9c7a`), so
  this used to work.

It looks like there's a mismatch between what `gitolite info` reports
and what's actually enforced at push time. Could you take a look at
the ACL for `packages/TSSr` when you have a moment?

If it's easier on your end, I'm happy to send a tarball or a patch
bundle instead.

Thanks for the help!
Junhao Chen, PhD

**Comment by lshep** [comment]
Git.bioconductor.org does not have a master branch you need to push to devel with something like git push upstream master:devel

**Comment by JohnnyChen1113** [comment]
> Dear Package contributor,
> 
> This is the automated single package builder at bioconductor.org.
> 
> Your package has been built on the Bioconductor Single Package Builder.
> 
> On one or more platforms, the build results were: "WARNINGS". This may mean there is a problem with the package that you need to fix. Or it may mean that there is a problem with the build system itself.
> 
> Please see the [build report](http://bioconductor.org/spb_reports/TSSr_buildreport_20260507181405.html) for more details.
> 
> The following are build products from R CMD build on the Single Package Builder: Linux (Ubuntu 24.04.4 LTS): [TSSr_0.99.15.tar.gz](https://bioconductor.org/spb_reports/TSSr_0.99.15.tar.gz)
> 
> Links above active for 21 days.
> 
> ** Remember: **if you submitted your package after July 7th, 2020, when making changes to your repository push to `git@git.bioconductor.org:packages/TSSr` to trigger a new build. A quick tutorial for setting up remotes and pushing to upstream can be found [here](https://contributions.bioconductor.org/git-version-control.html#new-package-workflow).

 Hi @lshep,

The 0.99.15 build returns Status OK with one WARNING: "R CMD check exceeded 10 min requirement" (CheckTime: 11:10). All other checks pass:

 - 0 ERROR / 0 WARNING / 0 NOTES from BiocCheckGitClone
 - 0 ERROR / 0 WARNING / 4 NOTES from BiocCheck (only the standard "consider biocViews", function-length, line-length, indentation suggestions)
 - testthat: 0 FAIL / 0 SKIP / 154 PASS (skip helpers do not fire on SPB workers — IS_BIOC_BUILD_MACHINE is unset and our multi-signal helper also did not match)

 Where the time goes:
 - tests/testthat: 206s (the package's actual test suite, not benchmark/stress tests)
 - examples: 113s (exportClustersToBed alone is 51s — exporting consensus clusters to BED is the function's core purpose)
 - 6× package load/unload checks: ~95s (driven by Bioconductor genome/annotation imports such as BSgenome, GenomicFeatures, txdbmaker)
 - "checking R code for possible problems": 51s

 We've already wrapped pipeline-heavy examples in slot-access form to keep them runnable but fast (no \donttest{} on any exported man page). The remaining time is fundamentally tied to the package's domain (TSS-seq pipeline + genome dependencies) rather than test/example bloat.

 Could you advise whether this 11:10 check time is acceptable for review purposes, or whether further trimming is required? If so, would moving the slowest tests behind a skip_on_bioc()-equivalent gate (or shrinking exampleTSSr to a smaller fixture) be the preferred path?

 Thanks!
Junhao Chen, PhD

**Comment by hpages** [comment]
Sorry for the slow response.

The major issue with the package is that it does everything it can to avoid a fair QA assessment by our automated daily builds. It does this by:
  - Having man pages with examples that do not call the function documented in the man page.
  - Using `eval=FALSE` for the chunks in the vignette where these functions are called.
  - Using `.skip_on_bioc_spb()` in the unit tests where these functions are called so the functions won't actually get tested on the Bioconductor build/test machines.

This is the case for example for functions `clusterTSS()`, `annotateCluster()`, `callEnhancer()`, `consensusCluster()`, `deGene()`, `shapeCluster()`, and `shiftPromoter()`. All of them correspond to important steps in a typical **TSSr** workflow, but none of them gets _effectively_ used anywhere in the man pages or vignette, or will get _effectively_ tested on the Bioconductor build/test machines.

Guaranteeing the proper functioning of the packages that we host and distribute is at the very core of Bioconductor mission. But it only works if package authors see the value of this model and accept to play along, not if they meticulously take steps to defeat it.

Other issues:

13. Unit tests take more than 15 min to complete on my machine. This is too long. Unit tests should be quick whether they run on a Bioconductor build/test machine or not. The way to achieve this is either by using smaller data for the tests, or by optimizing your functions, or both.

14. Many hard dependencies: `library(TSSr)` takes 11.7 s on my laptop and loads 126 packages. Are you sure you need them all? Maybe some of the packages in `Imports` could be moved to `Suggests`? For example, **Gviz**, **DESeq2**, **ggfortify**, and **calibrate** don't seem to be needed for the core TSS-calling/clustering workflow, so maybe could be moved to `Suggests` and used via `if (requireNamespace())` conditionals? (This is the traditional way to handle suggested packages in functions that rely on them.)
  Note that I already reported this issue in my review from Feb 2 (this was item 12).

15. Many functions are implemented in a way that seems to achieve in-place modification of an object. For example:
    ```r
    mergeSamples(myTSSr)
    ```
    seems to modify `myTSSr` in-place.
    This is a very unconventional approach that will confuse most users because `myTSSr` is a classic S4 object, not a reference object. This means that the object is always passed by value, that is, the function that receives it is guaranteed to always modify _a copy of it_, never the original object.
    However you managed to simulate a pass-by-reference semantic by implementing the following hack:
    - put line `objName <- deparse(substitute(object))` at the beginning of the function body;
    - and put line `assign(objName, object, envir = parent.frame())` at the end of the function body.
    
    What this hack does is reassign the modified copy of the object to the original symbol of the object.
    
    This might seem clever, but it's not a good idea for various reasons. One of them is that the hack is not reliable -- it will fail in some situations. For example, if `mergeSamples()` was _really_ capable of doing _true_ in-place modification of the object, then the following code should also be able to do it:
    ```r
    toto <- function(x) mergeSamples(x)
    toto(myTSSr)
    ```
    But it doesn't.
    The bottom line is: the standard way to use a function like `mergeSamples()` in R/Bioconductor is like this:
    ```r
    myTSSr <- mergeSamples(myTSSr)
    ```
    This means that your functions simply need to return the modified object. That's all!

16. Displaying a TSSr object (e.g. object `exampleTSSr`) produces many screens of output. TSSr objects need a `show()` method that produces a display that is more compact than what we get at the moment.

17. Using curly brackets around the call to `standardGeneric()` in the definition of your S4 generic functions is problematic:
    ```r
    > clusterTSS
    nonstandardGenericFunction for "clusterTSS" defined from package "TSSr"

    function (object, method = "peakclu", peakDistance = 100, extensionDistance = 30, 
        localThreshold = 0.02, clusterThreshold = 1, useMultiCore = FALSE, 
        numCores = NULL) 
    {
        standardGeneric("clusterTSS")
    }
    <bytecode: 0x5ffca803a828>
    <environment: 0x5ffca02726d0>
    Methods may be defined for arguments: object, method, peakDistance, extensionDistance, localThreshold, clusterThreshold, useMultiCore, numCores
    Use  showMethods(clusterTSS)  for currently available ones.
    ```
    As you can see, doing so has the effect of making your function a `nonstandardGenericFunction` instead of a `standardGenericFunction`, which is almost certainly not what you want here.
  
18. Another issue with the definition of your S4 generic functions is that you're not specifying the signature for dispatch. As a consequence, all the arguments are involved in method dispatch which is usually not recommended.

19. A Bioconductor package should always be installed with `BiocManager::install()` and almost never directly from GitHub. Please fix the Installation section in your vignette.

Thanks,
H.

**Comment by JohnnyChen1113** [comment]
> Sorry for the slow response.
> 
> The major issue with the package is that it does everything it can to avoid a fair QA assessment by our automated daily builds. It does this by:
> 
> * Having man pages with examples that do not call the function documented in the man page.
> * Using `eval=FALSE` for the chunks in the vignette where these functions are called.
> * Using `.skip_on_bioc_spb()` in the unit tests where these functions are called so the functions won't actually get tested on the Bioconductor build/test machines.
> 
> This is the case for example for functions `clusterTSS()`, `annotateCluster()`, `callEnhancer()`, `consensusCluster()`, `deGene()`, `shapeCluster()`, and `shiftPromoter()`. All of them correspond to important steps in a typical **TSSr** workflow, but none of them gets _effectively_ used anywhere in the man pages or vignette, or will get _effectively_ tested on the Bioconductor build/test machines.
> 
> Guaranteeing the proper functioning of the packages that we host and distribute is at the very core of Bioconductor mission. But it only works if package authors see the value of this model and accept to play along, not if they meticulously take steps to defeat it.
> 
> Other issues:
> 
> 13. Unit tests take more than 15 min to complete on my machine. This is too long. Unit tests should be quick whether they run on a Bioconductor build/test machine or not. The way to achieve this is either by using smaller data for the tests, or by optimizing your functions, or both.
> 14. Many hard dependencies: `library(TSSr)` takes 11.7 s on my laptop and loads 126 packages. Are you sure you need them all? Maybe some of the packages in `Imports` could be moved to `Suggests`? For example, **Gviz**, **DESeq2**, **ggfortify**, and **calibrate** don't seem to be needed for the core TSS-calling/clustering workflow, so maybe could be moved to `Suggests` and used via `if (requireNamespace())` conditionals? (This is the traditional way to handle suggested packages in functions that rely on them.)
>     Note that I already reported this issue in my review from Feb 2 (this was item 12).
> 15. Many functions are implemented in a way that seems to achieve in-place modification of an object. For example:
>     mergeSamples(myTSSr)
>         
>           
>         
>     
>           
>         
>     
>         
>       
>     seems to modify `myTSSr` in-place.
>     This is a very unconventional approach that will confuse most users because `myTSSr` is a classic S4 object, not a reference object. This means that the object is always passed by value, that is, the function that receives it is guaranteed to always modify _a copy of it_, never the original object.
>     However you managed to simulate a pass-by-reference semantic by implementing the following hack:
>     
>     * put line `objName <- deparse(substitute(object))` at the beginning of the function body;
>     * and put line `assign(objName, object, envir = parent.frame())` at the end of the function body.
>     
>     What this hack does is reassign the modified copy of the object to the original symbol of the object.
>     This might seem clever, but it's not a good idea for various reasons. One of them is that the hack is not reliable -- it will fail in some situations. For example, if `mergeSamples()` was _really_ capable of doing _true_ in-place modification of the object, then the following code should also be able to do it:
>     toto <- function(x) mergeSamples(x)
>     toto(myTSSr)
>         
>           
>         
>     
>           
>         
>     
>         
>       
>     But it doesn't.
>     The bottom line is: the standard way to use a function like `mergeSamples()` in R/Bioconductor is like this:
>     myTSSr <- mergeSamples(myTSSr)
>         
>           
>         
>     
>           
>         
>     
>         
>       
>     This means that your functions simply need to return the modified object. That's all!
> 16. Displaying a TSSr object (e.g. object `exampleTSSr`) produces many screens of output. TSSr objects need a `show()` method that produces a display that is more compact than what we get at the moment.
> 17. Using curly brackets around the call to `standardGeneric()` in the definition of your S4 generic functions is problematic:
>     > clusterTSS
>     nonstandardGenericFunction for "clusterTSS" defined from package "TSSr"
>     
>     function (object, method = "peakclu", peakDistance = 100, extensionDistance = 30, 
>         localThreshold = 0.02, clusterThreshold = 1, useMultiCore = FALSE, 
>         numCores = NULL) 
>     {
>         standardGeneric("clusterTSS")
>     }
>     <bytecode: 0x5ffca803a828>
>     <environment: 0x5ffca02726d0>
>     Methods may be defined for arguments: object, method, peakDistance, extensionDistance, localThreshold, clusterThreshold, useMultiCore, numCores
>     Use  showMethods(clusterTSS)  for currently available ones.
>         
>           
>         
>     
>           
>         
>     
>         
>       
>     As you can see, doing so has the effect of making your function a `nonstandardGenericFunction` instead of a `standardGenericFunction`, which is almost certainly not what you want here.
> 18. Another issue with the definition of your S4 generic functions is that you're not specifying the signature for dispatch. As a consequence, all the arguments are involved in method dispatch which is usually not recommended.
> 19. A Bioconductor package should always be installed with `BiocManager::install()` and almost never directly from GitHub. Please fix the Installation section in your vignette.
> 
> Thanks, H.

Dear @hpages ,

Thank you for the detailed review. We also want to apologize directly for the way the previous submission handled Bioconductor QA. The package contained a helper specifically designed to detect Single Package Builder workers and skip the core workflow tests there. That was wrong and inconsistent with the purpose of Bioconductor's automated QA.

We have deleted that helper, all of its call sites, the equivalent inline logic, and every other Bioconductor-build-specific skip condition. We did not replace them with `skip_on_bioc()`, environment checks, `\donttest{}`, `\dontrun{}`, or another bypass. The examples, tests, and vignette now execute the real workflow on Bioconductor workers.

The revised package is version **0.99.19**. The current GitHub source is [commit 9fbd9b5](https://github.com/Linlab-slu/TSSr/commit/9fbd9b5b29fb285e8d0ea7a916bc118430bcc19b).

## SPB status

The same commit was pushed to the Bioconductor `devel` branch on July 31. The issue bot acknowledged the push at 18:45 UTC and reported that it was starting a build for commit `9fbd9b5b29fb285e8d0ea7a916bc118430bcc19b`. As of August 3, however, no build completion message or report has been posted. The public SPB history endpoint at `staging.bioconductor.org:8000` also currently returns an empty response and resets the connection for us.

We have not repeated the push or bumped the version again, because we do not want to create duplicate build jobs. Could you please let us know whether this build is still queued, or whether it needs to be restarted?

## Core QA changes

- The examples for `getTSS()`, `clusterTSS()`, `consensusCluster()`, `annotateCluster()`, `shapeCluster()`, `deGene()`, `callEnhancer()`, and `shiftPromoter()` now call the documented function and use its returned object.
- The vignette executes the complete workflow. `getTSS()` imports a bundled, reproducible 100-row, four-sample TSStable fixture with 50 records from each strand, and no workflow chunk uses `eval=FALSE`.
- Public-API integration tests cover all six documented `getTSS()` input types and both the `refTable` and GFF annotation paths using small, reproducible bundled fixtures.
- The complete tests still run when we set the same build-machine signals used by the deleted helper:

  ```text
  IS_BIOC_BUILD_MACHINE=true BBS_HOME=/fake/bbs
  FAIL 0 | WARN 75 | SKIP 0 | PASS 992
  ```

- Repository-wide searches now find no Bioconductor-specific skip helper, `skip_on_bioc()`, `skip_if*()`, `\donttest{}`, or `\dontrun{}`.
- We removed legacy tests that only exercised dependency behavior rather than TSSr. Plot/export prerequisites and single-sample consensus output are now asserted explicitly, so an empty result cannot silently produce a PASS.

The 75 testthat warnings are retained rather than hidden: 42 are chi-squared-approximation warnings from sparse promoter-shift tables, 32 report non-positive observations that cannot be drawn on logarithmic axes, and one is an upstream ggfortify `aes_string()` deprecation warning. These do not produce an `R CMD check` warning.

## 13. Test duration

The original complete testthat run took 918 seconds. We reduced the bundled example data to chromosome I, regenerated every downstream result by executing the public workflow, and removed a duplicate full-pipeline test after moving its unique assertions into the main workflow tests.

The reduced object contains 29,456 raw TSS rows and 125 reference genes and is reproducibly built by `inst/scripts/make-exampleTSSr.R`. Repeated complete test runs now take approximately 1--3 minutes depending on machine load, with zero skips.

## 14. Hard dependencies

We moved DESeq2, Gviz, ggfortify, and calibrate from `Imports` to `Suggests`. Their feature entry points use `requireNamespace()` with an actionable installation message, and their calls are namespace-qualified.

Using fresh R sessions with the same measurement procedure, newly loaded namespaces decreased from 127 to 87, a reduction of 40 namespaces (31.5%). None of these four optional packages is loaded by `library(TSSr)`.

## 15. Simulated in-place modification

We removed all 12 occurrences of the `deparse(substitute())` plus `assign(..., parent.frame())` pattern from the 11 workflow methods. Each method now visibly returns the modified `TSSr` object. Internal calls, examples, tests, the README, and the vignette now consistently use standard R value semantics:

```r
myTSSr <- mergeSamples(myTSSr)
```

Regression tests verify both that the returned object contains the result and that an unassigned call does not modify the caller's object. They also cover the wrapper scenario described in the review.

## 16. Compact display

TSSr now has a documented and exported `show()` method. It reports object dimensions and completed stages without printing entire slots. Empty objects, long sample lists, side-effect freedom, and `print()`/`show()` consistency are covered by tests.

## 17--18. S4 generics and dispatch

All 25 original exported generics are now runtime `standardGeneric` objects. We removed the braces around `standardGeneric()` and explicitly set `signature = "object"`. Contract tests verify their classes, signatures, formals, and default values.

We interpreted these newer, specific instructions as superseding the broader December 2021 suggestion to replace all S4 generics with ordinary functions. The 13 read-only accessors added below follow the same standard-generic and object-only-dispatch requirements.

## 19. Installation instructions

The README and vignette now present the normal installation route as:

```r
BiocManager::install("TSSr")
```

GitHub installation is identified only as an option for the development version.

## Earlier review items

We re-audited all comments from the February 2026 and 2021 reviews. In particular:

- `makeTxDbFromGFF()` is called through txdbmaker. We removed an unnecessary taxonomy lookup from the temporary TxDb construction, so GFF annotation no longer depends on GenomeInfoDbData or the spelling/capitalization of an organism name. `organismName` remains optional descriptive metadata for backward compatibility; annotation uses the coordinates and strands in `refSource` or `refTable`, as in the published workflow.
- `sapply()` and unsafe `1:n` patterns were removed; progress output uses `message()`; normal assignments use `<-`.
- All 19 man pages specifically identified in the earlier review contain runnable examples that call their own documented function.
- `TSSr()` is a documented constructor with input checks, and the class has a validity method.
- We chose the review's graceful-error option for the original BAM files that cannot be bundled with `exampleTSSr`. Missing files now produce an informative error. Separately, a reproducible 100-row, four-sample TSStable fixture allows the `getTSS()` example and vignette to perform a real import.
- The `exampleTSSr` Format section now describes the S4 object correctly; generated `docs/` and `Meta/` directories are not tracked.
- rmarkdown and the yeast BSgenome are in `Suggests`; the unused GenomicAlignments dependency was removed.
- `callEnhancer()` now documents its returned `TSSr` object accurately.

We also resolved the concern about exposing internal data.table objects. TSSr now provides 13 documented read-only accessors covering its matrices, reference data, library sizes, and analysis results. Every table accessor returns an independent base `data.frame` copy. All examples, the README, and the vignette use these accessors rather than direct `slot()` or `@` access.

## Final local validation

```text
Package version:                 0.99.19
R CMD build:                    OK
R CMD check --no-manual:        Status: OK
R CMD check diagnostics:        0 ERROR | 0 WARNING | 0 NOTE
Examples and vignette rebuild:  OK
Tests:                          FAIL 0 | WARN 75 | SKIP 0 | PASS 992
BiocCheck 1.48.0:               0 ERROR | 0 WARNING | 5 NOTE
```

The BiocCheck notes are non-blocking recommendations concerning a suggested biocView, existing function length and formatting, and the automatic mailing list subscription check. The maintainer confirms that `zhenguo.lin@slu.edu` is subscribed to bioc-devel.

We recognize that the previous QA approach damaged confidence in the submission. We have therefore tried not only to remove the bypass, but also to make the workflow reproducible and add regression tests for the issues uncovered by executing it.

Thank you again for the review and for insisting that the package execute its real workflow during QA. We would also appreciate any guidance on the pending SPB job described above.

**Comment by hpages** [comment]
Thanks for all the changes. It sounds like we are going in the right direction.

Regarding the pending SPB job: It looks like we finally have a build report (see above), 4 days after commit `9fbd9b5b29fb285e8d0ea7a916bc118430bcc19b`! No idea why it took so long. This is VERY unusual.

As you can see in the report, a test is failing. Please address, bump the package version, commit and push.

I'll take a closer look at all the recent changes once the package passes BUILD and CHECK with no error.

Thanks again,
H.

**Comment by JohnnyChen1113** [comment]
> Thanks for all the changes. It sounds like we are going in the right direction.
> 
> Regarding the pending SPB job: It looks like we finally have a build report (see above), 4 days after commit `9fbd9b5b29fb285e8d0ea7a916bc118430bcc19b`! No idea why it took so long. This is VERY unusual.
> 
> As you can see in the report, a test is failing. Please address, bump the package version, commit and push.
> 
> I'll take a closer look at all the recent changes once the package passes BUILD and CHECK with no error.
> 
> Thanks again, H.

Dear @hpages,

  Thank you. We addressed the failing test, bumped the package version to 0.99.20, and pushed commit [`7c60c71`](https://github.com/Linlab-slu/TSSr/
  commit/7c60c71cd64f4619165220389dd0038b1c61b8fb) to the Bioconductor `devel` branch.

  The failure was caused by a platform-dependent floating-point boundary case in the 10th/90th percentile calculation. A cumulative signal mathematically equal to the threshold could fall on
  opposite sides of a strict comparison on macOS and Linux. We replaced this with a scale-aware, inclusive threshold rule and added regression tests covering the exact-threshold case and
  positive rescaling.

  The new SPB report is clean:

  - BUILD: OK
  - R CMD check: Status OK
  - POST-PROCESSING: OK
  - BiocCheck: 0 errors, 0 warnings, and 4 advisory notes
  - SPB check time: 7 minutes 10 seconds

  Report: [TSSr 0.99.20 SPB build report](http://bioconductor.org/spb_reports/TSSr_buildreport_20260806134929.html)

  Thank you again for your patience and for reviewing these changes. The package is now ready for your closer review, and we will promptly address anything else you identify.

Best,
Junhao

**Comment by hpages** [comment]
Thanks for all your hard work on **TSSr**. I think we're almost there.

A couple more things and we should be good to go:

### Vignette

The vignette is a little bit too dry. It's a succession of code chunks without much context. It would be nice to provide a little bit more explanations/details about what's going on by putting one or two sentences before and/or after each code chunk.

The chi-squared-approximation warnings introduce a lot of noise and might confuse the user:

<img width="756" height="1052" alt="Image" src="https://github.com/user-attachments/assets/b0baac30-9244-4999-9089-c5f07556e2da" />

In particular the fact that the same warning is repeated so many times is problematic. Is there a way that `shiftPromoter()` could be modified to display the warning only once? Also the user might not necesarily understand why they get the warning so please explain the meaning of the warning and whether it can safely be ignored or not.  The best place for such explanation is the man page of the `shiftPromoter()` function and the vignette could just refer the reader to that.

Please make sure that the text of the vignette is in sync with the content of the code chunks. For example at the beginning of section "TSS data processing" you refer to the `exampleTSSr` object, but then you use the `myTSSr` object in the code chunk. This can only confuse the reader. Maybe the two objects are the same but it doesn't matter: users who jump directly to this section won't necessarily know about this and will be really confused.

### Mandatory vs optional arguments

There should be a clear distinction between mandatory arguments and optional arguments. Right now this distinction is very blurry.

For example, all the arguments of the `TSSr()` constructor function have default values. As a consequence, the user can call the function without specifying any argument. Surprisingly this returns an object without complaining, even though nothing will work with this object.

Only optional arguments should have default values.

Another more realistic situation where this causes problems is the `TSSr()` call that you show in the vignette to create the `importedTSSr` object. The code that you show does not specify the `refSource` argument. Is this an optional argument? Hard to know. On one hand the argument has a default value (`character()`) but on the other hand the man page for `TSSr()` doesn't say that the argument is optional. Note that it says so for the `organismName` argument but not for `refSource`.

But then trying to perform the `annotateCluster` step fails with the following error:
```r
annotateCluster(importedTSSr)

# Annotating...
# Error in annotateCluster(importedTSSr) : Please provide correct refSource file!
```
`TSSr()` shouldn't let the user create "incomplete" objects that will fail them down the road. At least not without saying anything. Well-designed software checks user input and fails early rather than late.

Also why would `TSSr()` let the user specify a file that doesn't exist at creation time?
```r
importedTSSr <- TSSr(
    genomeName = "BSgenome.Scerevisiae.UCSC.sacCer3",
    inputFiles = exampleInput,
    inputFilesType = "TSStable",
    sampleLabels = c("SL01", "SL02", "SL03", "SL04"),
    sampleLabelsMerged = c("control", "treat"),
    mergeIndex = c(1, 1, 2, 2),
    refSource = "saccharomyces_cerevisiae.SGD.gff"
)

...

...

annotateCluster(importedTSSr)

# Annotating...
# Import genomic features from the file as a GRanges object ... Error in open.connection(con, open) : cannot open the connection
# In addition: Warning message:
# In open.connection(con, open) :
#   cannot open file 'saccharomyces_cerevisiae.SGD.gff': No such file or directory
```

Again, user input should be checked early rather than late.

My final attempt at fixing `importedTSSr` didn't succeed either:
```r
refSource <- system.file("extdata", "example-annotation.gff3", package="TSSr")
importedTSSr <- TSSr(
    genomeName = "BSgenome.Scerevisiae.UCSC.sacCer3",
    inputFiles = exampleInput,
    inputFilesType = "TSStable",
    sampleLabels = c("SL01", "SL02", "SL03", "SL04"),
    sampleLabelsMerged = c("control", "treat"),
    mergeIndex = c(1, 1, 2, 2),
    refSource = refSource
)

annotateCluster(importedTSSr)

# Annotating...
# Import genomic features from the file as a GRanges object ... OK
# Prepare the 'metadata' data frame ... OK
# Make the TxDb object ... OK
# Error in `[.data.table`(new, , .SD, .SDcols = filteredColumns) :
#   Some items of .SDcols are not column names: [cluster, chr, start, end, strand, dominant_tss, tags, tags.dominant_tss, q_0.1, q_0.9, ...]
# In addition: Warning message:
# In .makeTxDb_normarg_chrominfo(chrominfo) :
#   genome version information is not available for this TxDb object
```

So a major concern here is this: even though everything seems to run smootly on the `myTSSr` object used in the vignette, how is this going to work when people try to use this on their own data? This is the question that one should keep in mind at all time when designing software that is meant to be used by others.

Best,
H.

**Comment by JohnnyChen1113** [comment]
> Dear Package contributor,
> 
> This is the automated single package builder at bioconductor.org.
> 
> Your package has been built on the Bioconductor Single Package Builder.
> 
> Congratulations! The package built without errors or warnings on all platforms.
> 
> Please see the [build report](http://bioconductor.org/spb_reports/TSSr_buildreport_20260831101332.html) for more details.
> 
> The following are build products from R CMD build on the Single Package Builder: Linux (Ubuntu 24.04.4 LTS): [TSSr_0.99.21.tar.gz](https://bioconductor.org/spb_reports/TSSr_0.99.21.tar.gz)
> 
> Links above active for 21 days.
> 
> ** Remember: **if you submitted your package after July 7th, 2020, when making changes to your repository push to `git@git.bioconductor.org:packages/TSSr` to trigger a new build. A quick tutorial for setting up remotes and pushing to upstream can be found [here](https://contributions.bioconductor.org/git-version-control.html#new-package-workflow).

Dear @hpages,

  Thank you for the detailed review. We addressed both groups of issues, bumped the package version to **0.99.21**, and pushed commit [`bba29dd`](https://github.com/Linlab-slu/TSSr/commit/bba29ddb31aa3bcc5c8e1a4276224de759b5b7ea) to the Bioconductor `devel` branch.

  ### Vignette and promoter-shift warnings

  - We revised the vignette into an end-to-end workflow using the bundled TSStable input and GFF annotation. Explanatory context is now provided around each analysis stage and its output.
  - The workflow consistently uses `myTSSr`. The precomputed `exampleTSSr` object is now introduced separately and is not used as the workflow input.
  - `shiftPromoter()` now consolidates repeated chi-squared approximation warnings into one summary warning after all gene-level tests.
  - The `shiftPromoter()` man page and vignette explain that small expected counts can make the approximate p values unreliable and that the warning should not be ignored for inferential use.

  ### Mandatory and optional arguments

  - `genomeName`, `inputFiles`, `inputFilesType`, and `sampleLabels` are now required arguments of `TSSr()`. Calling `TSSr()` without them fails immediately.
  - Every input path must identify an existing regular file at construction time, and `getTSS()` revalidates the paths immediately before import.
  - `sampleLabelsMerged` and `mergeIndex` are optional but must be supplied together. When both are omitted, each sample forms its own group.
  - `refSource` remains optional because annotation is an optional workflow stage. Its documentation now states this explicitly and explains that `annotateCluster()` requires either `refSource` or an existing reference table. When supplied, `refSource` must identify exactly one existing regular file.

  The vignette now exercises the scenario from your review: it constructs an object from the bundled TSStable and GFF files and runs the complete workflow through `annotateCluster()`. A permanent integration test covers the same path. We also fixed the empty promoter/coding-overlap case that caused the internal `.SDcols` error, added workflow-stage and chromosome-name validation, and added focused annotation regression tests.

  The new SPB report is clean:

  - BUILD: OK
  - R CMD check: Status OK
  - Examples and tests: OK
  - BiocCheck: 0 errors, 0 warnings, and 4 advisory notes
  - SPB check time: 7 minutes 27 seconds

  Report: [TSSr 0.99.21 SPB build report](http://bioconductor.org/spb_reports/TSSr_buildreport_20260831101332.html)

  Thank you again for the careful review. Please let us know if any further changes are needed.

  Best,
  Junhao

**Comment by hpages** [comment]
Thanks for the changes. Package is good to go.

**Comment by lshep** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/Linlab-slu.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("TSSr")`. The package 'landing page' will be created at

https://bioconductor.org/packages/TSSr

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
